api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java

```diff
@@ -827,21 +815,185 @@ private String generateRandomEightCharacterQualifier() {
         .collect(Collectors.toList());
   }
 
+  private List<String> generateWgsCode(
+      DataSetExportRequest dataSetExportRequest, DbWorkspace dbWorkspace, String qualifier) {
+    if (!dataSetExportRequest.getGenerateGenomicsAnalysisCode()) {
+      return new ArrayList<>();
+    }
+
+    if (!workbenchConfigProvider.get().featureFlags.enableGenomicExtraction) {
+      throw new NotImplementedException();
+    }
+
+    if (Strings.isNullOrEmpty(dbWorkspace.getCdrVersion().getWgsBigqueryDataset())) {
+      throw new FailedPreconditionException(
+          "The workspace CDR version does not have whole genome data");
+    }
+
+    if (dataSetExportRequest.getKernelType().equals(KernelTypeEnum.R)) {
+      return generateGenomicsAnalysisComment_R();
+    }
+
+    switch (dataSetExportRequest.getGenomicsAnalysisTool()) {
+      case HAIL:
+        return generateHailCode(qualifier, dataSetExportRequest);
+      case PLINK:
+        return generatePlinkCode(qualifier, dataSetExportRequest);
+      case NONE:
+        return generateDownloadVcfCode(qualifier, dataSetExportRequest);
+      default:
+        throw new BadRequestException("Invalid Genomics Analysis Tool");
+    }
+  }
+
+  private Optional<String> getExtractionDirectory(Long datasetId) {
+    try {
+      return Optional.of(
+              submissionDao
+                  .findMostRecentValidExtractionByDataset(dataSetDao.findById(datasetId).get())
+                  .get()
+                  .getOutputDir()
+                  .replaceFirst("/$", "")) // Drop trailing slash if exists
+          .filter(dir -> !dir.isEmpty());
+    } catch (NoSuchElementException e) {
+      return Optional.empty();
+    }
+  }
+
+  private String generateVcfDirEnvName(String qualifier) {
+    return "DATASET_" + qualifier + "_VCF_DIR";
+  }
+
+  private static final String MISSING_EXTRACTION_DIR_PLACEHOLDER =
+      "\"GCS_VCF_DIRECTORY_GOES_HERE\"";
+
+  private String generateExtractionDirCode(
+      String qualifier, DataSetExportRequest dataSetExportRequest) {
+    String extractionDir =
+        getExtractionDirectory(dataSetExportRequest.getDataSetRequest().getDataSetId())
+            .orElse(MISSING_EXTRACTION_DIR_PLACEHOLDER);
+
+    String noExtractionDirComment =
+        MISSING_EXTRACTION_DIR_PLACEHOLDER.equals(extractionDir)
+            ? "# VCF files for this dataset do not exist\n"
+                + "# Run a Genomic Extraction from a Dataset to generate VCF files\n"
+            : "";
+
+    return noExtractionDirComment
+        + "%env "
+        + generateVcfDirEnvName(qualifier)
+        + "="
+        + extractionDir;
+  }
+
+  private List<String> generateHailCode(
+      String qualifier, DataSetExportRequest dataSetExportRequest) {
+    final String matrixName = "mt_" + qualifier;
+
+    return ImmutableList.of(
+        "# Note: Hail demos must be run from a \"Hail Genomics Analysis\" cloud analysis environment",
```

Comment:
I'd suggest adding this to the "Initialize Hail" code cell instead, I don't think it warrants its own cell.